### DISCUSSION TITLE
How to detect whether JRuby app is running on Tomcat?

### DISCUSSION THREAD

**Discussion by tedt10e** [jruby/jruby#6657] [Q&A]
Hello @headius 

I'm not sure where to ask, but I think you could help us shed some light on the canonical way to detect whether the app is serving from Tomcat / Java Application Server / Standalone Server.

This is related to the PR for good_job gems https://github.com/bensheldon/good_job/pull/230

Many Thanks 🙏

Ted

**Comment by bensheldon** [comment]
Thanks @tedhexaflow and @headius!

Background: GoodJob is a Rails ActiveJob backend. GoodJob can run threaded within the Rails process. 

The challenge: I'd like to detect whether GoodJob is running within a "Webserver Process" (Webrick, Puma, Passenger, Tomcat, Servlet, etc)  or any other processes (rails generate..., rake db:migrate..., etc).  

In MRI... I've circled in on using `Rails.const_defined?('Server')` and `caller.grep(%r{/config.ru}).any?` (detect whether `config.ru` is in the callstack). Based on this [SO Post](https://stackoverflow.com/a/49598328) and confirmed experimentally.

In JRuby, I'm a little stuck due to my own unfamiliarity.  I'm packaging with Warbler, which uses jruby-rack (Are there other build tools I should support?). It seems like the resulting packages can be run in a few different ways:

- Directly as a webserver `java -jar app.war`
- Using Tomcat (or another Java webserver)
- Directly as an executable `java -jar app.war -S rake db:migrate`

Solutions?

- Directly as a webserver `java -jar app.war`. It seems like ``uri:classloader:/jruby/rack/rails_booter.rb:83:in `load_environment'`` is reliably in the callstack, meaning I can use something like 
    ```ruby
       caller.grep(%r{jruby/rack/rails_booter}).any?
    ```
- Using Tomcat (or another Java webserver). Seems like the previous would work also because Tomcat is invoking the servlet.
- Directly as an executable `java -jar app.war -S rake db:migrate`. I'm not sure and have had trouble getting this to work.
- Other strategies I should consider?

**Reply by headius** [reply]
Hmmm ok this is an interesting one. My first thought would be that all the server processes you list would at some level be using Rack to handle requests. Even the Tomcat and other WAR file deployments will work that way, because they use the jruby-rack library.

Is looking for Rack good enough? Alternatively, if you want to know you are actually processing a web request, I'm sure there are constants or environment variables that Rack sets up for you.

**Reply by bensheldon** [reply]
@headius That's really helpful to know!

I think I have a working set of logic described at https://github.com/bensheldon/good_job/pull/230#issuecomment-817950264

> Is looking for Rack good enough?

Unfortunately, after experimentation, it seems like Rack is likely to be initialized during Rails initialization even if the webserver itself isn't being used. 

> ...if you want to know you are actually processing a web request...

I'm trying to fingerprint this during Rails initialization outside of a request.
